datafusion/functions/src/regex/regexpsubstr.rs

```diff
@@ -0,0 +1,531 @@
+            });
+
+        let is_scalar = len.is_none();
+        let inferred_length = len.unwrap_or(1);
+        let args = args
+            .args
+            .iter()
+            .map(|arg| arg.to_array(inferred_length))
+            .collect::<Result<Vec<_>>>()?;
+
+        let result = regexp_match_func(&args);
+        if is_scalar {
+            // If all inputs are scalar, keeps output as scalar
+            let result = result.and_then(|arr| ScalarValue::try_from_array(&arr, 0));
+            result.map(ColumnarValue::Scalar)
+        } else {
+            result.map(ColumnarValue::Array)
+        }
+    }
+
+    fn documentation(&self) -> Option<&Documentation> {
+        Some(get_regexp_substr_doc())
+    }
+}
+
+static DOCUMENTATION: OnceLock<Documentation> = OnceLock::new();
+
+fn get_regexp_substr_doc() -> &'static Documentation {
+    DOCUMENTATION.get_or_init(|| {
+        Documentation::builder(
+            DOC_SECTION_REGEX,
+            "Returns the substring that matches a [regular expression](https://docs.rs/regex/latest/regex/#syntax) within a string.",
+            "regexp_substr(str, regexp[, position[, occurrence[, flags[, group_num]]]])")
+            .with_sql_example(r#"```sql
+            > select regexp_substr('Köln', '[a-zA-Z]ö[a-zA-Z]{2}');
+            +---------------------------------------------------------+
+            | regexp_substr(Utf8("Köln"),Utf8("[a-zA-Z]ö[a-zA-Z]{2}")) |
+            +---------------------------------------------------------+
+            | [Köln]                                                  |
+            +---------------------------------------------------------+
+            SELECT regexp_substr('aBc', '(b|d)', 1, 1, 'i');
+            +---------------------------------------------------+
+            | regexp_substr(Utf8("aBc"),Utf8("(b|d)"), Int32(1), Int32(1), Utf8("i")) |
+            +---------------------------------------------------+
+            | [B]                                               |
+            +---------------------------------------------------+
+```
+Additional examples can be found [here](https://docs.snowflake.com/en/sql-reference/functions/regexp_substr#examples)
+"#)
+            .with_standard_argument("str", Some("String"))
+            .with_argument("regexp", "Regular expression to match against.
+            Can be a constant, column, or function.")
+            .with_argument("position", "Number of characters from the beginning of the string where the function starts searching for matches. Default: 1")
+            .with_argument("occurrence", "Specifies the first occurrence of the pattern from which to start returning matches.. Default: 1")
+            .with_argument("flags",
+                           r#"Optional regular expression flags that control the behavior of the regular expression. The following flags are supported:
+  - **i**: case-insensitive: letters match both upper and lower case
```

Comment:
I believe we want to support Snowflake-like regex_parameters, not DF-like. They are not exactly the same. I see 2 options here:
1. Upstream this with DF-style regex-flags. So, it will be more likely to be accepted. Then implement additional Snowflake flags only for our fork as a separate task.
2. Try to gamble and make Snowflake flags in this PR, and then we'll see what happens on the external review.